### ISSUE TITLE
S3 service prototype missing createBucket function

### ISSUE THREAD

**Issue by Ghazgkull** [aws/aws-sdk-js#3412]
Confirm by changing [ ] to [x] below to ensure that it's a bug:
- [x] I've gone through [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/welcome.html) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/)
- [x] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js) for answers
- [x] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js/issues) and didn't find any solution
  
**Describe the bug**

The prototype configured for S3 service instances is missing the createBucket function.

**Is the issue in the browser/Node.js?**
Node.js

**If on Node.js, are you running this on AWS Lambda?**

No.

**Details of the browser/Node.js version**
Paste output of `npx envinfo --browsers` or `node -v`

Node v12.18.3

**SDK version number**

2.724.0

**To Reproduce (observed behavior)**

1. Instantiate an instance of the S3 service (i.e. `new S3()`)
2. Get the prototype of the service instance (i.e. `Object.getPrototypeOf(service)`)
3. Inspect the properties on the prototype. At first glance, it appears to contain all of the API functions. But on closer inspection, the `createBucket` function is missing.

```
const { S3 } = require('aws-sdk');

const getExportedFunctions = (object) => {
  return Object.getOwnPropertyNames(Object.getPrototypeOf(object))
    .filter((name) => name !== 'constructor' && typeof object[name] === 'function')
    .sort();
};

const instance = new S3();
const exported = getExportedFunctions(instance);
console.log(`exported: ${JSON.stringify(exported)}`);
```

This code outputs the following list of functions:
```
"abortMultipartUpload",
"completeMultipartUpload",
"copyObject",
"createMultipartUpload",
"deleteBucket",
"deleteBucketAnalyticsConfiguration",
"deleteBucketCors",
"deleteBucketEncryption",
"deleteBucketInventoryConfiguration",
"deleteBucketLifecycle",
"deleteBucketMetricsConfiguration",
"deleteBucketPolicy",
"deleteBucketReplication",
"deleteBucketTagging",
"deleteBucketWebsite",
"deleteObject",
"deleteObjectTagging",
"deleteObjects",
"deletePublicAccessBlock",
"getBucketAccelerateConfiguration",
"getBucketAcl",
"getBucketAnalyticsConfiguration",
"getBucketCors",
"getBucketEncryption",
"getBucketInventoryConfiguration",
"getBucketLifecycle",
"getBucketLifecycleConfiguration",
"getBucketLocation",
"getBucketLogging",
"getBucketMetricsConfiguration",
"getBucketNotification",
"getBucketNotificationConfiguration",
"getBucketPolicy",
"getBucketPolicyStatus",
"getBucketReplication",
"getBucketRequestPayment",
"getBucketTagging",
"getBucketVersioning",
"getBucketWebsite",
"getObject",
"getObjectAcl",
"getObjectLegalHold",
"getObjectLockConfiguration",
"getObjectRetention",
"getObjectTagging",
"getObjectTorrent",
"getPublicAccessBlock",
"headBucket",
"headObject",
"listBucketAnalyticsConfigurations",
"listBucketInventoryConfigurations",
"listBucketMetricsConfigurations",
"listBuckets",
"listMultipartUploads",
"listObjectVersions",
"listObjects",
"listObjectsV2",
"listParts",
"putBucketAccelerateConfiguration",
"putBucketAcl",
"putBucketAnalyticsConfiguration",
"putBucketCors",
"putBucketEncryption",
"putBucketInventoryConfiguration",
"putBucketLifecycle",
"putBucketLifecycleConfiguration",
"putBucketLogging",
"putBucketMetricsConfiguration",
"putBucketNotification",
"putBucketNotificationConfiguration",
"putBucketPolicy",
"putBucketReplication",
"putBucketRequestPayment",
"putBucketTagging",
"putBucketVersioning",
"putBucketWebsite",
"putObject",
"putObjectAcl",
"putObjectLegalHold",
"putObjectLockConfiguration",
"putObjectRetention",
"putObjectTagging",
"putPublicAccessBlock",
"restoreObject",
"selectObjectContent",
"uploadPart",
"uploadPartCopy"
```

**Expected behavior**

The createBucket function should be included on the prototype.

**Additional context**

This is relevant to me because I am trying to wrap the S3 service instance in a [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy), for the purpose of doing my own enhanced logging and tracing.

**Comment by AllanFly120** [comment]
Hi @Ghazgkull 

This is expected. JS SDK contains handwritten customization code to overwrite the default behavior of `createBucket`. You can find the implementation from [the source code](https://github.com/aws/aws-sdk-js/blob/5ae5a7d7d24d1000dbc089cc15f8ed2c7b06c542/lib/services/s3.js#L1327-L1346). As a result, unlike other methods that can be found in `Object.getPrototypeOf(s3)` or `s3.__proto__`, the `createBucket()` is defined in `s3.__proto__.__proto__`. 

If you want to get all the naive methods  s3 client, you can print `Object.keys(s3.api.operations)`

**Comment by Ghazgkull** [comment]
Thanks, @AllanFly120.

I have tried using new S3().api.operations. This includes the `createBucket` function as you mentioned, but unfortunately it's missing a number of other API functions. Specifically:
```"createPresignedPost","getSignedUrl","getSignedUrlPromise","upload","waitFor"```

**Comment by Ghazgkull** [comment]
I tried four different approaches in order to programatically get the list of available service functions:
1. Inspecting the prototype properties as illustrated here.
  - This was missing the createBucket function as reported.
2. Walking the prototype chain with a slightly modified version of this code.
  - This yielded the same result.
3. Loading the JSON API descriptors bundled with the aws-sdk. i.e. ('aws-sdk/apis/s3-2006-03-01.min.json')
  - This was actually pretty good, but it doesn't work for the DynamoDB.DocumentClient because no such API descriptor is provided.
4. Inspecting the serviceInstance.api.operations property.
  - This is also missing a number of functions as mentioned in my previous comment.

At the end of the day, in order to really get an accurate list of service functions, I had to resort to opening Chrome devtools on the API documentation website and writing a quick `querySelectorAll()` selector to capture all the function names. Then I took those names and captured them in my own JSON file per service that I bundle with my code. Super gross, but it works and was the only solution I could find.
